src/nvim/ui.h

```diff
@@ -22,6 +23,7 @@ struct ui_t {
   void (*clear)(UI *ui);
   void (*eol_clear)(UI *ui);
   void (*cursor_goto)(UI *ui, int row, int col);
+  void (*cursor_style_set)(UI *ui, Dictionary args); // or a Dictionary ?
```

Comment:
I understood that correctly but your sugggestion in https://github.com/neovim/neovim/issues/2583#issuecomment-272988384 made me wonder.
I can only think of pushing each cursorentry via several calls of :
 
```
void (*cursor_style_set)(UI *ui,   int shape;                    /* one of the SHAPE_ defines */
  int mshape,                   /* one of the MSHAPE defines */
  int percentage,               /* percentage of cell for bar */
  long blinkwait,               /* blinking, wait time before blinking starts */
  long blinkon,                 /* blinking, on time */
  long blinkoff,                /* blinking, off time */
  int id,                       /* highlight group ID */
  int id_lm,                    /* highlight group ID for :lmap mode */
  char        *name,            /* mode name (fixed) */
char used_for, /* SHAPE_MOUSE and/or SHAPE_CURSOR */)
```